ndonnx/_typed_array/onnx.py

```diff
@@ -2633,3 +2639,86 @@ def numeric_like(x: TyArrayBase) -> TyArrayInt32:
+    # ort_compat.py, but doing the below in spox-operators is just
+    # needless self-inflicted pain.
+
+    # The core idea of this mitigation is to do a runtime check on the
+    # dimensions of the inputs and the then hide the matmul inside an
+    # if-subgraph. However, there are a couple of complications:
+
+    #   1. The output of the zero-size branch has to match the shape of
+    # the output of the regular matmul. The semantics of that shape
+    # operation are involved
+    # (https://data-apis.org/array-api/draft/API_specification/generated/array_api.matmul.html#array_api.matmul).
+
+    #   2. We have to disable the value propagation or else onnxruntime
+    # will throw a segfault when executing the matmul branch (which is
+    # always happening).
+
+    # We grind our way through 1. The second point is solved by creating
+    # a small ONNX model and inlining it. We could consider doing this
+    # in a more generic fashion to enable faster build times.
+
+    # TODO: Fix upstream in onnxruntime!
+    def dummy_n_1(a: TY_ARRAY_NUMBER, b: TY_ARRAY_NUMBER) -> TyArrayInt64:
+        return a.dynamic_shape[:-1]
+
+    def dummy_1_n(a: TY_ARRAY_NUMBER, b: TY_ARRAY_NUMBER) -> TyArrayInt64:
+        out_shape = b.dynamic_shape
+        out_shape = out_shape[:-2].concat([out_shape[-1:]], axis=0)
+        return out_shape
+
+    def dummy_n_m(a: TY_ARRAY_NUMBER, b: TY_ARRAY_NUMBER) -> TyArrayInt64:
+        a_shape = a.dynamic_shape
+        dummy_a_shape = a_shape.copy()
+        dummy_a_shape[-2:] = const(1)
+
+        b_shape = b.dynamic_shape
+        dummy_b_shape = b_shape.copy()
+        dummy_b_shape[-2:] = const(1)
+
+        dummy_a = zeros(shape=dummy_a_shape, dtype=uint8)
+        dummy_b = zeros(shape=dummy_b_shape, dtype=uint8)
+
+        out_shape = (dummy_a + dummy_b).dynamic_shape
+        out_shape[-2] = a_shape[-2]
+        out_shape[-1] = b_shape[-1]
+        return out_shape
+
+    a_ = type(a)(argument(a._var.unwrap_type()))
+    b_ = type(b)(argument(b._var.unwrap_type()))
+
+    if a_.ndim == 1 and b_.ndim == 1:
+        out_shape = const(np.asarray([], np.int64), int64)
+    elif a_.ndim >= 2 and b_.ndim == 1:
+        out_shape = dummy_n_1(a_, b_)
+    elif a_.ndim == 1 and b_.ndim >= 2:
+        out_shape = dummy_1_n(a_, b_)
+    elif a_.ndim >= 2 and b_.ndim >= 2:
+        out_shape = dummy_n_m(a_, b_)
+    else:
+        raise ValueError(
+            "unsupported input ranks for 'matmul': `{a_.ndim}` and `{b_.ndim}`"
+        )
+
+    dummy_out = zeros(shape=out_shape, dtype=a.dtype)._var
+    res = op.matmul(a_._var, b_._var)
+
+    (var,) = op.if_(
+        safe_cast(TyArrayBool, (a_.dynamic_size * b_.dynamic_size == 0))._var,
+        then_branch=lambda: [dummy_out],
+        else_branch=lambda: [res],
+    )
+    model = build({"a": a_._var, "b": b_._var}, {"res": var})
+    (var_with_propagation,) = inline(model)(a=a._var, b=b._var).values()
+
+    arr = type(a)(var_with_propagation)
```

Comment:
This section seems more appropriate for `ort_compat` for guarding against zero dimensions. I get why the previous bit would stay here.

Reply:
But the previous part is needed to compute `dummy_out`. I'm not quite seeing how we would be able to nicely move this part into `ort_compat.py`

Reply:
I was imaging providing dummy and result branches as zero arg lambdas and then doing something like `opx.guard_zero_dims(evaluate=lambda: ..., fallback=lambda:...)`.

Reply:
Thus far everything in `ort_compat` just mirrors standard operators. Ideally, we would have a `ort_compat.matmul` that Does The Right Thing. Then, once we have things patched upstream we can just remove the workaround in `ort_compat` and profit everywhere without touching any code elsewhere. However, I'm not sure how we would use the proposed `guard_zero_dims` in this context. What we could do, and what may actually be nicer, is that we do move this function into `ort_compat` but use lazy imports to make it work. What do you think of that?